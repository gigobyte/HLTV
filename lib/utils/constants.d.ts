export declare const HLTV_URL = "http://www.hltv.org";
export declare const HLTV_STATIC_URL = "https://static.hltv.org/";
