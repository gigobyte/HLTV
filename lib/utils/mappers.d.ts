import Team from '../models/Team';
import Veto from '../models/Veto';
import Player from '../models/Player';
import { MapStatistic } from '../models/FullTeam';
import { WeakRoundOutcome } from '../models/RoundOutcome';
import MapSlug from '../enums/MapSlug';
export declare const fetchPage: (url: string) => Promise<any>;
export declare const toArray: (elements: any) => any[];
export declare const getMapSlug: (map: string) => MapSlug;
export declare const mapVetoElementToModel: (el: any, team1: Team, team2: Team) => Veto;
export declare const getMatchPlayer: (playerEl: any) => Player;
export declare const getMatchFormatAndMap: (mapText: string) => {
    map?: MapSlug | undefined;
    format: string;
};
export declare const mapRoundElementToModel: (team1Id: number, team2Id: number) => (el: any, i: number) => WeakRoundOutcome;
export declare const getMapsStatistics: (source: string) => {
    [key: string]: MapStatistic;
} | undefined;
export declare const getTimestamp: (source: string) => number;
