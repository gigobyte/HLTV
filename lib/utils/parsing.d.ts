export declare const text: (el: any) => any;
export declare const prev: (el: any) => any;
export declare const hasChild: (childSelector: string) => (el: any) => boolean;
export declare const popSlashSource: (el: any) => any;
