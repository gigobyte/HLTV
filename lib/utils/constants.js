"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HLTV_URL = 'http://www.hltv.org';
exports.HLTV_STATIC_URL = 'https://static.hltv.org/';
