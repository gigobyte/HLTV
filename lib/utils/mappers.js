"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var cheerio = require("cheerio");
var RoundOutcome_1 = require("../models/RoundOutcome");
var MapSlug_1 = require("../enums/MapSlug");
var E = require("../utils/parsing");
var rp = require('request-promise');
var Agent = require('socks5-http-client/lib/Agent');
exports.fetchPage = function (url) { return __awaiter(_this, void 0, void 0, function () {
    var _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                _b = (_a = cheerio).load;
                return [4, rp.get({
                        url: url,
                        agentClass: Agent,
                        agentOptions: {
                            socksHost: process.env.SOCKET_HOST,
                            socksPort: process.env.SOCKET_PORT
                        }
                    })
                        .then(function (res) { return res.text(); })];
            case 1: return [2, _b.apply(_a, [_c.sent()])];
        }
    });
}); };
exports.toArray = function (elements) { return elements.toArray().map(cheerio); };
exports.getMapSlug = function (map) { return MapSlug_1.default[map]; };
exports.mapVetoElementToModel = function (el, team1, team2) {
    var _a = el.text().replace(/^\d. /, '').split(/removed|picked/), teamName = _a[0], map = _a[1];
    if (!map || !teamName) {
        return {
            map: exports.getMapSlug(el.text().split(' ')[1]),
            type: 'other'
        };
    }
    return {
        team: [team1, team2].find(function (t) { return t.name === teamName.trim(); }),
        map: exports.getMapSlug(map.trim()),
        type: el.text().includes('picked') ? 'picked' : 'removed'
    };
};
exports.getMatchPlayer = function (playerEl) {
    var link = playerEl.parent().attr('href');
    return {
        name: playerEl.find('.text-ellipsis').text(),
        id: link ? Number(link.split('/')[2]) : undefined
    };
};
exports.getMatchFormatAndMap = function (mapText) {
    if (mapText && !mapText.includes('bo')) {
        return { map: mapText, format: 'bo1' };
    }
    return { format: mapText };
};
exports.mapRoundElementToModel = function (team1Id, team2Id) { return function (el, i) {
    var outcomeString = E.popSlashSource(el).split('.')[0];
    var outcome = Object.entries(RoundOutcome_1.Outcome).find(function (_a) {
        var _ = _a[0], v = _a[1];
        return v === outcomeString;
    });
    return {
        outcome: outcome && outcome[1],
        score: el.attr('title'),
        ctTeam: i < 15 ? team1Id : team2Id,
        tTeam: i < 15 ? team2Id : team1Id
    };
}; };
exports.getMapsStatistics = function (source) {
    var valueRegex = /"value":"(\d|\.)+%?"/g;
    var labelRegex = /label":"\w+/g;
    var splitRegex = /%?"/;
    if (!source.match(labelRegex)) {
        return;
    }
    var maps = source.match(labelRegex).map(function (x) { return x.split(':"')[1]; });
    var values = source.match(valueRegex).map(function (x) { return Number(x.split(':"')[1].split(splitRegex)[0]); });
    var getStats = function (index) { return ({
        winningPercentage: values[index],
        ctWinningPercentage: values[index + maps.length * 1],
        tWinningPercentage: values[index + maps.length * 2],
        timesPlayed: values[index + maps.length * 3]
    }); };
    return maps.reduce(function (stats, map, i) {
        return (__assign({}, stats, (_a = {}, _a[MapSlug_1.default[map]] = getStats(i), _a)));
        var _a;
    }, {});
};
exports.getTimestamp = function (source) {
    var _a = source.split('/'), day = _a[0], month = _a[1], year = _a[2];
    return new Date([month, day, year].join('/')).getTime();
};
