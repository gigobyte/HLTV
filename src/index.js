import cheerio from 'cheerio'
import fetch from 'isomorphic-fetch'

const HLTV_URL = 'http://www.hltv.org'

class ParsingTools {
    _cleanupString(str) {
        if(str.includes('[emailÂ protected]')) {
            console.warn('Cannot parse element because of HLTV obfuscation')
        } else {
            return str.replace(/\s\s+/g, ' ').trim()
        }
    }

    _getTeamId($team) {
        const teamLink = $team.attr('href')

        if(teamLink && teamLink !== '#') return parseInt(teamLink.split('=')[2])
    }

    _getMatchLink($matchAction) {
        const matchLink = $matchAction.attr('href')

        if(matchLink) return matchLink.replace('/match/', '')
    }

    //DRY?
    _getMatchId(matchLink) {
        if(matchLink) return parseInt(matchLink.split('-')[0])
    }

    _parseMatchPageMaps($, $maps, $results) {
        let maps = Array.apply(null, Array($maps.length)).map(e => ({}))

        $maps.each((i, map) => {
            maps[i].map = $(map).find('img').attr('src').split('hotmatch/')[1].split('.png')[0]
        })

        $results.each((i, res) => {
            maps[i].result = this._cleanupString($(res).text())
        })

        return maps
    }

    _parseMatchPageStreams($, $streams) {
        return $streams.map((i, e) => ({
            name: this._cleanupString($(e).text()),
            link: HLTV_URL + $(e).find('a').attr('href')
        })).get()
    }

    _parseMatchPagePlayer($) {
        return (i, e) => this._cleanupString($(e).children().first().text())
    }

    _restructureMatch(match) {
        if(['LIVE', 'Finished'].includes(match.time)) {
            delete match.time
        }

        if(!(match.format.includes('Best of'))) {
            match.map = match.format
            match.format = 'Best of 1'
        }

        if(match.label) {
            delete match.team1
            delete match.team1Id
            delete match.team2
            delete match.team2Id
            delete match.live
            delete match.finished
            delete match.map
        } else {
            delete match.label
        }

        if(!match.team1Id) delete match.team1Id
        if(!match.team2Id) delete match.team2Id
    }

    _restructureFullMatch(match) {
        if(match.title) {
            delete match.team1
            delete match.team1Id
            delete match.team2
            delete match.team2Id
            delete match.highlights
            delete match.players
        } else {
            delete match.title
        }
    }
}

class HLTV extends ParsingTools {
    async getMatches() {
        let matches = []
        const response = await fetch(`${HLTV_URL}/matches/`).then(res => res.text())
        const $ = cheerio.load(response)
        const $matchElems = $('.matchListRow')

        $matchElems.each((i, elem) => {
            const $elem = $(elem)
            const $team1 = $elem.find('.matchTeam1Cell > a')
            const $team2 = $elem.find('.matchTeam2Cell > a')
            const $liveInfo = $($elem.find('.matchScoreCell > div > div'))

            let match = {}

            match.time     = $elem.find('.matchTimeCell').text()
            match.eventId  = parseInt($elem.parent().attr('class').split(' ')[1].replace('event_', ''))
            match.team1    = $team1.text().trim()
            match.team2    = $team2.text().trim()
            match.team1Id  = this._getTeamId($team1)
            match.team2Id  = this._getTeamId($team2)
            match.live     = (match.matchTime === 'LIVE')
            match.finished = (match.matchTime === 'Finished')
            match.format   = $($liveInfo[0]).text().trim()
            match.label    = $elem.find('div[style="text-align: center;width: 80%;float: left;"]').text()
            match.link     = this._getMatchLink($elem.find('.matchActionCell > a'))
            match.id       = this._getMatchId(match.link)

            this._restructureMatch(match)

            matches.push({...match})
        })

        return matches
    }

    async getLatestResults({pages = 1} = {}) {
        if(pages < 1) throw new Error('HLTV.getLatestResults: pages cannot be less than 1')

        let matches = []
        for(let i = 0; i < pages; i++) {
            const response = await fetch(`${HLTV_URL}/results/${i*50}/`).then(res => res.text())
            const $ = cheerio.load(response)
            const $matchElems = $('.matchListRow')

            $matchElems.each((i, elem) => {
                const $elem = $(elem)
                const $team1 = $elem.find('.matchTeam1Cell > a')
                const $team2 = $elem.find('.matchTeam2Cell > a')

                let match = {}

                match.format  = $elem.find('.matchTimeCell').text()
                match.team1   = $team1.text().trim()
                match.team2   = $team2.text().trim()
                match.team1Id = this._getTeamId($team1)
                match.team2Id = this._getTeamId($team2)
                match.link    = this._getMatchLink($elem.find('.matchActionCell > a'))
                match.id       = this._getMatchId(match.link)
                match.result  = $elem.find('.matchScoreCell').text().trim()

                this._restructureMatch(match)

                matches.push({...match})
            })
        }

        return matches
    }

    async getStreams({loadLinks = false} = {}) {
        let streams = []
        const response = await fetch(HLTV_URL).then(res => res.text())
        const $ = cheerio.load(response)
        const $streamTitles = $('div[style*="width: 95px;"]')
        const $streamViewers = $('div[style*="width: 35px;"]')

        for(let i = 0; i < $streamTitles.length; i++) {
            let stream = {}

            const $streamHref = $($streamTitles[i]).find('a')

            stream.name     = $streamHref.attr('title')
            stream.viewers  = parseInt($($streamViewers[i]).text().replace(/[()]/g, ''))
            stream.category = $($streamHref.find('img')[0]).attr('title')
            stream.country  = $($streamHref.find('img')[1]).attr('src').split('flag/')[1].split('.')[0]
            stream.hltvLink = HLTV_URL + $streamHref.attr('href')

            if(loadLinks) {
                const hltvPage = await fetch(stream.hltvLink).then(res => res.text())

                stream.realLink = cheerio.load(hltvPage)('iframe').attr('src')
            }

            streams.push({...stream})
        }

        return streams
    }

    async getMatch({id} = {}) {
        let match = {
            event: {},
            players: []
        }

        const response = await fetch(`${HLTV_URL}/match/${id}`).then(res => res.text())
        const $ = cheerio.load(response)

        const $teams = $('div[style*="width:46%;"]')
        const $eventInfo = $('div[style*="font-size: 18px;"]')
        const $mapFormatBox = $('#mapformatbox')
        const $maps = $('div[style*="width:280px;"]')
        const $mapResults = $('div[style*="width:270px;"]')
        const $highlights = $('.hotmatchroundbox').has('div[style="cursor:pointer;color:#0269D2"]')
        const $streams = $('.hotmatchroundbox').has('div[style="cursor:pointer;width:240px;"]')
        const $players = $('div[style*="width:105px;"]')
        const $playerHighlight = $('.headertext').find('a > b')

        const $team1 = $($teams[0]).find('span > a')
        const $team2 = $($teams[1]).find('span > a')

        match.team1 = $team1.text().trim()
        match.team2 = $team2.text().trim()
        match.team1Id = this._getTeamId($team1)
        match.team2Id = this._getTeamId($team2)
        match.date = this._cleanupString($($eventInfo[0]).text())

        match.event.name = $eventInfo.find('a').text()
        match.event.link = HLTV_URL + $eventInfo.find('a').attr('href')
        match.format = $mapFormatBox.text().split('\n')[1].trim()

        match.maps = this._parseMatchPageMaps($, $maps, $mapResults)
        match.streams = this._parseMatchPageStreams($, $streams)
        match.highlights = $highlights.map((i, e) => this._cleanupString($(e).text())).get()

        match.players[0] = $players.slice(0,5).map(this._parseMatchPagePlayer($)).get()
        match.players[1] = $players.slice(5,10).map(this._parseMatchPagePlayer($)).get()

        if($mapFormatBox.text().split('\n').length > 3) {
            match.additionalInfo = $mapFormatBox.text().split('\n')[3].trim()
        }

        if($playerHighlight.text()) {
            match.playerHighlight = {
                playerName: $playerHighlight.text().slice(1, -1),
                playerId: parseInt($playerHighlight.parent().attr('href').split('playerid=')[1])
            }
        }

        match.title = this._cleanupString($('span[style*="font-size: 26px"] > span').text())

        this._restructureFullMatch(match)

        return match
    }
}

export default HLTV
