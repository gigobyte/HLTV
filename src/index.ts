import { defaultConfig, defaultLoadPage, HLTVConfig } from './config'
import { getMatch } from './endpoints/getMatch'
import { connectToScorebot } from './endpoints/connectToScorebot'
import { getMatches } from './endpoints/getMatches'
import { getEvent } from './endpoints/getEvent'
import { getEventByName } from './endpoints/getEventByName'
import { getEvents } from './endpoints/getEvents'
import { getMatchMapStats } from './endpoints/getMatchMapStats'
import { getMatchStats } from './endpoints/getMatchStats'
import { getMatchesStats } from './endpoints/getMatchesStats'
import { getPlayer } from './endpoints/getPlayer'
import { getPlayerByName } from './endpoints/getPlayerByName'
import { getPlayerRanking } from './endpoints/getPlayerRanking'
import { getPlayerStats } from './endpoints/getPlayerStats'
import { getRecentThreads } from './endpoints/getRecentThreads'
import { getStreams } from './endpoints/getStreams'
import { getTeam } from './endpoints/getTeam'
import { getTeamByName } from './endpoints/getTeamByName'
import { getTeamRanking } from './endpoints/getTeamRanking'
import { getTeamStats } from './endpoints/getTeamStats'
import { getPastEvents } from './endpoints/getPastEvents'
import { getResults } from './endpoints/getResults'
import { getNews } from './endpoints/getNews'

export class Hltv {
  constructor(private config: Partial<HLTVConfig> = {}) {
    if (config.httpAgent && !config.loadPage) {
      config.loadPage = defaultLoadPage(config.httpAgent)
    }

    if (!config.httpAgent) {
      config.httpAgent = defaultConfig.httpAgent
    }

    if (!config.loadPage) {
      config.loadPage = defaultConfig.loadPage
    }
  }

  getMatch = getMatch(this.config as HLTVConfig)
  getMatches = getMatches(this.config as HLTVConfig)
  getEvent = getEvent(this.config as HLTVConfig)
  getEvents = getEvents(this.config as HLTVConfig)
  getPastEvents = getPastEvents(this.config as HLTVConfig)
  getEventByName = getEventByName(this.config as HLTVConfig)
  getMatchMapStats = getMatchMapStats(this.config as HLTVConfig)
  getMatchStats = getMatchStats(this.config as HLTVConfig)
  getMatchesStats = getMatchesStats(this.config as HLTVConfig)
  getPlayer = getPlayer(this.config as HLTVConfig)
  getPlayerByName = getPlayerByName(this.config as HLTVConfig)
  getPlayerRanking = getPlayerRanking(this.config as HLTVConfig)
  getPlayerStats = getPlayerStats(this.config as HLTVConfig)
  getRecentThreads = getRecentThreads(this.config as HLTVConfig)
  getStreams = getStreams(this.config as HLTVConfig)
  getTeam = getTeam(this.config as HLTVConfig)
  getTeamByName = getTeamByName(this.config as HLTVConfig)
  getTeamRanking = getTeamRanking(this.config as HLTVConfig)
  getTeamStats = getTeamStats(this.config as HLTVConfig)
  getResults = getResults(this.config as HLTVConfig)
  getNews = getNews(this.config as HLTVConfig)
  connectToScorebot = connectToScorebot(this.config as HLTVConfig)

  public createInstance(config: Partial<HLTVConfig>) {
    return new Hltv(config)
  }

  public TEAM_PLACEHOLDER_IMAGE =
    'https://www.hltv.org/img/static/team/placeholder.svg'

  public PLAYER_PLACEHOLDER_IMAGE =
    'https://static.hltv.org/images/playerprofile/bodyshot/unknown.png'
}

const hltv = new Hltv()

export default hltv
export { hltv as HLTV }

export { MatchStatus } from './endpoints/getMatch'
export type {
  Demo,
  Highlight,
  Veto,
  HeadToHeadResult,
  ProviderOdds,
  MapHalfResult,
  MapResult,
  Stream,
  FullMatch as Match
} from './endpoints/getMatch'

export { MatchEventType, MatchFilter } from './endpoints/getMatches'
export type { MatchPreview, GetMatchesArguments } from './endpoints/getMatches'

export { WinType } from './endpoints/connectToScorebot'
export type { ScoreboardUpdate, LogUpdate } from './endpoints/connectToScorebot'

export type {
  FullEvent,
  FullEventHighlight,
  FullEventFormat,
  FullEventPrizeDistribution,
  FullEventTeam
} from './endpoints/getEvent'

export type { EventPreview, GetEventsArguments } from './endpoints/getEvents'

export type { FullMatchStats } from './endpoints/getMatchStats'

export type {
  GetMatchesStatsArguments,
  MatchStatsPreview
} from './endpoints/getMatchesStats'

export type {
  FullPlayerTeam,
  PlayerAchievement,
  FullPlayer
} from './endpoints/getPlayer'

export type {
  PlayerRanking,
  GetPlayerRankingOptions
} from './endpoints/getPlayerRanking'

export type {
  FullPlayerStats,
  GetPlayerStatsArguments
} from './endpoints/getPlayerStats'

export { ThreadCategory } from './endpoints/getRecentThreads'
export type { Thread } from './endpoints/getRecentThreads'

export { StreamCategory } from './endpoints/getStreams'
export type { FullStream } from './endpoints/getStreams'

export { TeamPlayerType } from './endpoints/getTeam'
export type { FullTeam, FullTeamPlayer } from './endpoints/getTeam'

export type { TeamRanking, GetTeamArguments } from './endpoints/getTeamRanking'

export type { GetPastEventsArguments } from './endpoints/getPastEvents'

export {
  ResultsMatchType,
  ContentFilter,
  GameType
} from './endpoints/getResults'
export type {
  FullMatchResult,
  ResultTeam,
  GetResultsArguments
} from './endpoints/getResults'

export type { NewsPreview, GetNewsArguments } from './endpoints/getNews'

export { GameMap } from './shared/GameMap'
export { MatchFormat } from './shared/MatchFormat'
export { RankingFilter } from './shared/RankingFilter'
export { MatchType } from './shared/MatchType'
export { BestOfFilter } from './shared/BestOfFilter'
export type { Article } from './shared/Article'
export type { Country } from './shared/Country'
export type { Event } from './shared/Event'
export type { Player } from './shared/Player'
export type { Team } from './shared/Team'
export type { EventType } from './shared/EventType'
