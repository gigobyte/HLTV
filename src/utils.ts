import * as cheerio from 'cheerio'
import { v4 as uuidv4 } from 'uuid'

export const fetchPage = async (
  url: string,
  loadPage: (url: string) => Promise<string>
): Promise<cheerio.Root> => {
  const root = cheerio.load(await loadPage(url))

  const html = root.html()

  if (
    html.includes('error code:') ||
    html.includes('Sorry, you have been blocked')
  ) {
    throw new Error(
      'Access denied | www.hltv.org used Cloudflare to restrict access'
    )
  }

  return root
}

export const generateRandomSuffix = () => {
  return uuidv4()
}

export const percentageToDecimalOdd = (odd: number): number =>
  parseFloat(((1 / odd) * 100).toFixed(2))

export function getIdAt(index: number, href: string): number | undefined
export function getIdAt(index: number): (href: string) => number | undefined
export function getIdAt(index?: number, href?: string): any {
  switch (arguments.length) {
    case 1:
      return (href: string) => getIdAt(index!, href)
    default:
      return parseNumber(href!.split('/')[index!])
  }
}

export const notNull = <T>(x: T | null): x is T => x !== null

export const parseNumber = (str: string | undefined): number | undefined => {
  if (!str) {
    return undefined
  }

  const num = Number(str)

  return Number.isNaN(num) ? undefined : num
}

export const sleep = (ms: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, ms))
}
