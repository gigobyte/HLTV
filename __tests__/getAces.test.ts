import HLTV, { GameMap } from '../src/'
import { sleep } from '../src/utils'

const EVENT_1 = 5605
const EVENT_2 = 6372
const MAPS = [GameMap.Overpass, GameMap.Nuke]

test('getAces', async () => {
  await sleep(3000)
  expect(await HLTV.getAces({ eventIds: [EVENT_1] })).toMatchSnapshot()
  await sleep(3000)
  expect(
    await HLTV.getAces({ eventIds: [EVENT_2], maps: MAPS })
  ).toMatchSnapshot()
  await sleep(3000)
}, 30000)
