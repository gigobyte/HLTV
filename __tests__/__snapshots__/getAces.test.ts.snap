// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAces 1`] = `
Array [
  Object {
    "date": "25/9/21",
    "map": "Inferno",
    "player": "rain",
    "round": "6",
    "team1": "FaZe",
    "team2": "OG",
  },
  Object {
    "date": "25/9/21",
    "map": "Inferno",
    "player": "flameZ",
    "round": "28",
    "team1": "FaZe",
    "team2": "OG",
  },
  Object {
    "date": "25/9/21",
    "map": "Ancient",
    "player": "Twistzz",
    "round": "16",
    "team1": "OG",
    "team2": "FaZe",
  },
  Object {
    "date": "25/9/21",
    "map": "Overpass",
    "player": "Perfecto",
    "round": "15",
    "team1": "Natus Vincere",
    "team2": "Complexity",
  },
  Object {
    "date": "24/9/21",
    "map": "Ancient",
    "player": "Twistzz",
    "round": "11",
    "team1": "Complexity",
    "team2": "FaZe",
  },
  Object {
    "date": "20/9/21",
    "map": "Inferno",
    "player": "REZ",
    "round": "2",
    "team1": "NIP",
    "team2": "BIG",
  },
  Object {
    "date": "20/9/21",
    "map": "Inferno",
    "player": "AMANEK",
    "round": "6",
    "team1": "MIBR",
    "team2": "G2",
  },
  Object {
    "date": "18/9/21",
    "map": "Vertigo",
    "player": "Bubzkji",
    "round": "3",
    "team1": "Astralis",
    "team2": "Vitality",
  },
  Object {
    "date": "18/9/21",
    "map": "Dust2",
    "player": "Kyojin",
    "round": "8",
    "team1": "Vitality",
    "team2": "Astralis",
  },
  Object {
    "date": "16/9/21",
    "map": "Inferno",
    "player": "Stewie2K",
    "round": "20",
    "team1": "Vitality",
    "team2": "Liquid",
  },
]
`;

exports[`getAces 2`] = `
Array [
  Object {
    "date": "17/5/22",
    "map": "Overpass",
    "player": "Magisk",
    "round": "9",
    "team1": "Vitality",
    "team2": "Heroic",
  },
  Object {
    "date": "15/5/22",
    "map": "Nuke",
    "player": "Ax1Le",
    "round": "22",
    "team1": "Cloud9",
    "team2": "FaZe",
  },
  Object {
    "date": "14/5/22",
    "map": "Nuke",
    "player": "cadiaN",
    "round": "1",
    "team1": "Heroic",
    "team2": "Spirit",
  },
]
`;
